File: Backtracking/Subsets.java
// LeetCode 78 - Subsets
// Time Complexity: O(n * 2^n) | Space Complexity: O(n)
import java.util.ArrayList;
import java.util.List;

public class Subsets {
    private final List<List<Integer>> answer = new ArrayList<>();
    private final List<Integer> current = new ArrayList<>();
    private int[] nums;

    public List<List<Integer>> subsets(int[] nums) {
        this.nums = nums;
        dfs(0);
        return answer;
    }

    private void dfs(int index) {
        if (index == nums.length) {
            answer.add(new ArrayList<>(current));
            return;
        }

        dfs(index + 1);

        current.add(nums[index]);
        dfs(index + 1);
        current.remove(current.size() - 1);
    }
}

File: Backtracking/Subsets.go
// LeetCode 78 - Subsets
// Time Complexity: O(n * 2^n) | Space Complexity: O(n)
package backtracking

func subsets(nums []int) (ans [][]int) {
	current := []int{}

	var dfs func(int)
	dfs = func(index int) {
		if index == len(nums) {
			ans = append(ans, append([]int(nil), current...))
			return
		}

		dfs(index + 1)
		current = append(current, nums[index])
		dfs(index + 1)
		current = current[:len(current)-1]
	}

	dfs(0)
	return
}
